Updated Date Range to Follow Documentation When Assuming Missing Values - Yaml Compat Doc Updates (#113964)
elastic/elasticsearch: REST_API_COMPATIBILITY.md

In some cases the prior version of the YAML REST tests are not sufficient to fully test changes. This can happen when the prior version has insufficient test coverage. In those cases, you can simply add more testing to the prior version or you can add custom REST tests that will run along side of the other compatibility tests. These custom tests can be found in the `yamlRestCompatTest` sourceset. Custom REST tests for compatibility will not be modified prior to execution, so the correct headers need to be manually added.

#### Breaking Changes

> It is possible to be in a state where you have intentionally made a breaking change and the compatibility tests will fail irrespective of checks for `skip` or `requires` cluster or test features in the current version. In this state, assuming the breaking changes are reasonable and agreed upon by the breaking change committee, the correct behavior is to skip the test in the `build.gradle`. For example, if you make a breaking change that causes the `range/20_synthetic_source/Date range` to break then this test can be disabled temporarily in this file `rest-api-spec/build.gradle` like within this snippet:

Author: same here I took this whole section and tried to better flesh out examples with versions let me know if that helped at all

```groovy
tasks.named("yamlRestCompatTestTransform").configure({task ->
    task.skipTest("range/20_synthetic_source/Date range", "date range breaking change causes tests to produce incorrect values for compatibility")
    task.skipTest("indices.sort/10_basic/Index Sort", "warning does not exist for compatibility")
    task.skipTest("search/330_fetch_fields/Test search rewrite", "warning does not exist for compatibility")
    task.skipTestsByFilePattern("indices.create/synthetic_source*.yml", "@UpdateForV9 -> tests do not pass after bumping API version to 9 [ES-9597]")
})
```

> Minimally once this skip test is in place, the correct `skip` and `requires` checks will need to be backported. Once backported the yaml compatibility tests should now work without the skip in the `build.gradle` file, and it should be removed.  Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes.

Author: tried to add in refs to 9.0.0 and 8.latest; let me know if that helps at all?

> Minimally once this skip test is in place, the correct `skip` and `requires` checks will need to be backported to say 8.latest. Once backported, the test can be re-enabled in 9.0.0 by removing the `skipTest`. Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes.  As an example in 9.0.0 a `requires` should be added for new tests to validate the changed api or output and `skip` added for the old tests which break in 9.0.0.  Then in backport say to 8.latest the `requires` check should be added to the existing tests with the old behavior so they no longer break when running bwc or upgrade tests from 9.0.0 to 8.latest.

Author: @andreidan completely fine with the rewrite here and agree it helps; updated to reflect your suggestion (with only minor changes to the wording).  Let me know if you feel good about that update and I'll wait a bit for any additional feedback and then merge.

> Minimally once this skip test is in place, the correct `skip` and `requires` checks will need to be backported to say 8.latest. Once backported, the test can be re-enabled in 9.0.0 by removing the `skipTest`. Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes.  As an example in 9.0.0 a `requires` should be added for new tests to validate the changed api or output and `skip` added for the old tests which break in 9.0.0.  Then in backport say to 8.latest the `requires` check should be added to the existing tests with the old behavior so they no longer break when running bwc or upgrade tests from 9.0.0 to 8.latest.

Reviewer: Thanks for iterating here. LGTM

> Minimally once this skip test is in place, the correct `skip` and `requires` checks will need to be backported to say 8.latest. Once backported the yaml compatibility tests should now work without the skip in the `build.gradle` file in 9.0.0, and it should be removed.  Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes.  As an example in 9.0.0 a `requires` should be added for new tests to validate the changed api or output and `skip` added for the old tests which break in 9.0.0.  Then in backport say to 8.latest the `requires` check should be added to the existing tests with the old behavior so they no longer break when running bwc or upgrade tests from 9.0.0 to 8.latest.

Reviewer: ```suggestion
Minimally once this skip test is in place, the correct `skip` and `requires` checks will need to be backported to say 8.latest. Once backported, the test can be re-enabled in 9.0.0 by removing the `skipTest`. Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes.  As an example in 9.0.0 a `requires` should be added for new tests to validate the changed api or output and `skip` added for the old tests which break in 9.0.0.  Then in backport say to 8.latest the `requires` check should be added to the existing tests with the old behavior so they no longer break when running bwc or upgrade tests from 9.0.0 to 8.latest.
```

Reviewer: >Consideration for whether tests should be cleaned up and changed based on how the breaking changes were backported is at the discretion of the team making the changes. 

Sorry this still reads a bit confusing to me. Namely I'm not sure what's optionally cleaned up here? 

> As an example in 9.0.0 a `requires` should be added for new tests to validate the changed api or output and `skip` added for the old tests which break in 9.0.0. Then in backport say to 8.latest the `requires` check should be added to the existing tests with the old behavior so they no longer break when running bwc or upgrade tests from 9.0.0 to 8.latest.

Can you please clarify where in the process this happens? The paragraph sort of concluded the backport above where we say "Once backported ...". Is this the same backport or a different flow altogether? 
Shall we add an actual example where we illustrate `requires` in `9.0.0` and `skip` in the backport branch?

#### Test Features

> Both cluster and test features exist. Cluster features are meant for new capability and test features can specifically be used to gate and manage `skip` and `requires` yaml test operations. Specifically in the context of backward compatibility and backporting, cluster and test features must be used consistently. When backporting and using these features they can not overlap in name and must be consistent when backported so that clusters built with these features are compatible.

Author: @andreidan honestly I don't know if we have documentation for that (if we do I haven't seen it).  And since test features are relatively new I'm not sure we have official guidance somewhere.  

@thecoop do you have more context there and/or additional docs we could link to?  What do you think about the current state of this paragraph?

I tried to add an additional sentence to provide more context here based on my understanding of the difference between the two.  Let me know what ya'll think.

> Both cluster and test features exist. Cluster features are meant for new capability and test features can specifically be used to gate and manage `skip` and `requires` yaml test operations. Specifically in the context of backward compatibility and backporting, cluster and test features must be used consistently. When backporting and using these features they can not overlap in name and must be consistent when backported so that clusters built with these features are compatible.

Reviewer 2: I think the paragraph is a bit confusing. Capabilities and features have different use cases. The docs in `docs/internal/Versioning.md` cover the different uses. I would just link to that doc, and only say here that capabilities and feature names need to be consistent across branches for the same functionality.

> Both cluster and test features exist. Cluster features are meant for new capability and test features can specifically be used to gate and manage `skip` and `requires` yaml test operations.  For more information, see `docs/internal/Versioning.md`. Specifically in the context of backward compatibility and backporting, cluster and test features must be used consistently across branches. When backporting and using these features they can not overlap in name and must be consistent when backported so that clusters built with these features are compatible.

Reviewer: Shall we link to `Versioning.md`? (there's an example [here](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc?plain=1#L64) on how to link )

> Both cluster and test features exist. Cluster features are meant for new capability and test features can specifically be used to gate and manage `skip` and `requires` yaml test operations.  For more information, see `docs/internal/Versioning.md`. Specifically in the context of backward compatibility and backporting, cluster and test features must be used consistently across branches. When backporting and using these features they can not overlap in name and must be consistent when backported so that clusters built with these features are compatible.

Reviewer: ```suggestion
Both cluster and test features exist. Cluster features are meant for new capability and test features can specifically be used to gate and manage `skip` and `requires` yaml test operations.  For more information, see `docs/internal/Versioning.md`. When backporting and using these features they can not overlap in name and must be consistent when backported so that clusters built with these features are compatible.
```
